build-system/pr-check/build.js

```diff
@@ -0,0 +1,116 @@
+const {determineBuildTargets} = require('./build-target');
+const {getStderr} = require('../exec');
+const {gitDiffColor} = require('../git');
+const {isTravisPullRequestBuild} = require('../travis');
+const FILENAME = 'build.js';
+const FILELOGPREFIX = colors.bold(colors.yellow(`${FILENAME}:`));
+const timedExecOrDie =
+  (cmd, unusedFileName) => timedExecOrDieBase(cmd, FILENAME);
+
+/**
+ * Makes sure package.json and yarn.lock are in sync.
+ * @private
+ */
+function runYarnIntegrityCheck_() {
+  const yarnIntegrityCheck = getStderr('yarn check --integrity').trim();
+  if (yarnIntegrityCheck.includes('error')) {
+    console.error(FILELOGPREFIX, colors.red('ERROR:'),
+        'Found the following', colors.cyan('yarn'), 'errors:\n' +
+        colors.cyan(yarnIntegrityCheck));
+    console.error(FILELOGPREFIX, colors.red('ERROR:'),
+        'Updates to', colors.cyan('package.json'),
+        'must be accompanied by a corresponding update to',
+        colors.cyan('yarn.lock'));
+    console.error(FILELOGPREFIX, colors.yellow('NOTE:'),
+        'To update', colors.cyan('yarn.lock'), 'after changing',
+        colors.cyan('package.json') + ',', 'run',
+        '"' + colors.cyan('yarn install') + '"',
+        'and include the updated', colors.cyan('yarn.lock'),
+        'in your PR.');
+    process.exit(1);
+  }
+}
+
+/**
+ * Makes sure that yarn.lock was properly updated.
+ * @private
+ */
+function runYarnLockfileCheck_() {
+  const localChanges = gitDiffColor();
+  if (localChanges.includes('yarn.lock')) {
+    console.error(FILELOGPREFIX, colors.red('ERROR:'),
+        'This PR did not properly update', colors.cyan('yarn.lock') + '.');
+    console.error(FILELOGPREFIX, colors.yellow('NOTE:'),
+        'To fix this, sync your branch to', colors.cyan('upstream/master') +
+        ', run', colors.cyan('gulp update-packages') +
+        ', and push a new commit containing the changes.');
+    console.error(FILELOGPREFIX, 'Expected changes:');
+    console.log(localChanges);
+    process.exit(1);
+  }
+}
+
+function main() {
+  const startTime = startTimer(FILENAME, FILENAME);
+
+  // Make sure package.json and yarn.lock are in sync and up-to-date.
+  runYarnIntegrityCheck_();
+  runYarnLockfileCheck_();
+  printChangeSummary(FILENAME);
+  const buildTargets = determineBuildTargets();
+
+  if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp update-packages');
+    timedExecOrDie('gulp build --fortesting');
+    uploadBuildOutput(FILENAME);
+  } else {
+    if (buildTargets.has('RUNTIME') ||
+        buildTargets.has('UNIT_TEST') ||
+        buildTargets.has('INTEGRATION_TEST') ||
+        buildTargets.has('BUILD_SYSTEM')) {
+
+      timedExecOrDie('gulp update-packages');
+      timedExecOrDie('gulp build --fortesting');
+      uploadBuildOutput(FILENAME);
+    } else {
+      console.log('Skipping build job because this commit does ' +
+       'not affect the runtime, build system, or test files');
```

Comment:
Before merging, sanity check all test jobs to make sure they are correctly handling the case when the build is skipped.

For example, I see that we're running `gulp test --nobuild` for the `DEV_DASHBOARD` case. Does it need `gulp build` to have been run?

Reply:
Done! Thanks, made a few adjustments.